signer/src/dkg/verification.rs

```diff
@@ -0,0 +1,766 @@
+        // current state, and then continue to process messages until there are
+        // no more state changes. This is necessary because the state machine
+        // can receive out-of-order messages, and we need to ensure that we
+        // process all messages that we can before returning, including handling
+        // state transitions.
+        while self.process_queued_messages()? {}
+
+        Ok(())
+    }
+
+    /// Validates that the sender of a message is a valid signer in the signing
+    /// set.
+    pub fn validate_sender(&self, signer_id: u32, sender: PublicKey) -> Result<(), Error> {
+        let config = self.coordinator.get_config();
+        let wsts: PublicKey = config
+            .signer_public_keys
+            .get(&signer_id)
+            .ok_or(Error::UnknownSender(sender))?
+            .try_into()
+            .map_err(|_| Error::InvalidPublicKeys)?;
+
+        if wsts != sender {
+            return Err(Error::SignerPublicKeyMismatch {
+                signer_id,
+                wsts: wsts.into(),
+                sender: sender.into(),
+            });
+        }
+
+        Ok(())
+    }
+
+    /// Checks whether or not the given public key is a known sender in the
+    /// signing set. Returns `true` if the sender is known, `false` otherwise.
+    /// Returns an error if the public key is invalid.
+    fn assert_is_known_sender(&self, sender: PublicKey) -> Result<(), Error> {
+        let is_known = self
+            .coordinator
+            .get_config()
+            .signer_public_keys
+            .values()
+            .any(|key| {
+                PublicKey::try_from(key)
+                    .map(|pub_key| pub_key == sender)
+                    .unwrap_or(false)
+            });
+
+        if !is_known {
+            return Err(Error::UnknownSender(sender));
+        }
+
+        Ok(())
+    }
+
+    /// Resets the [`StateMachine`] to its initial state, clearing all messages,
+    /// setting its creation time to the current time, its state to
+    /// [`State::Idle`] and also calling [`Coordinator::reset`] on the
+    /// [`FrostCoordinator`].
+    pub fn reset(&mut self) {
+        self.wsts_messages.clear();
+        self.created_at = Instant::now();
+        self.state = State::Idle;
+        self.coordinator.reset();
+    }
+
+    /// Gets the current state of the [`StateMachine`].
+    pub fn state(&self) -> &State {
+        &self.state
+    }
+
+    /// Gets the aggregate key that is being verified.
+    pub fn aggregate_key(&self) -> XOnlyPublicKey {
+        self.aggregate_key
+    }
+
+    /// Gets whether or not this [`StateMachine`] has expired.
+    fn is_expired(&self) -> bool {
+        match self.timeout {
+            None => false,
+            Some(timeout) => self.created_at.elapsed() > timeout,
+        }
+    }
+
+    /// Asserts that the [`StateMachine`] is in a state where it
+    /// can process messages.
+    fn assert_valid_state(&self) -> Result<(), Error> {
```

Comment:
Maybe rename this to `check_can_process_messages`?  `Success`, `Idle`, and `Signing` don't seem like "invalid" states.

Reply:
Went with `assert_valid_state_for_processing`: [75c78bc](https://github.com/stacks-network/sbtc/pull/1324/commits/75c78bc3c46868836857bc3750671ccbec4c83c3)